software/firmware/libraries/comm_protocol/examples/rover20_comms.py

```diff
@@ -0,0 +1,129 @@
+        self.port.write(bytes([ID]))
+        self.port.write(bytes([0x01]))
+        for i in address:
+            self.port.write(bytes([i]))
+        self.port.write(bytes([0xff]))
+        while(self.port.out_waiting):
+            time.sleep(0.0001)
+        data_list = []
+        for i in address:
+            if(self.data[ID-1][i] == 'float'):
+                while(self.port.in_waiting < 4): time.sleep(0.0001)
+                data_list.append(struct.unpack('f',self.port.read(4)))
+            elif(self.data[ID-1][i] == 'int'):
+                while(self.port.in_waiting < 2): time.sleep(0.0001)
+                data_list.append(struct.unpack('h',self.port.read(2)))
+            elif(self.data[ID-1][i] == 'long'):
+                while(self.port.in_waiting < 4): time.sleep(0.0001)
+                data_list.append(struct.unpack('i',self.port.read(4)))
+            elif(self.data[ID-1][i] == 'chr'):
+                while(self.port.in_waiting < 1): time.sleep(0.0001)
+                data_list.append(struct.unpack('c',self.port.read(1)))
+            while(self.port.in_waiting < 1): time.sleep(0.0001)
+            if(not(self.port.read(1) == b'\x00')):
+                print("Type Error when reading data from a slave\nThis will also cause a \"Bad acknowledge\" Error")
+                time.sleep(0.020)
+                self.port.reset_input_buffer()
+                break
+        sync = self.port.read(5)
+        if(not(sync == b'\xff\xff\xff\xff\xff')):
+            print("Bad acknowledge!!!!")
+        return(data_list)
+
+        
+    def write_multiple(self,ID,adr,data):
+        self.port.write(bytes([0x00]))
+        self.port.write(bytes([ID]))
+        self.port.write(bytes([0x03]))
+        for i in range(len(adr)):
+            self.port.write(bytes([adr[i]]))
+            if(isinstance(data[i], int) and data[i] >= -32768 and data[i] <= 32767 and self.data[ID-1][adr[i]] == 'int'):
+                self.port.write((data[i]).to_bytes(2, byteorder='big'))
+            elif(isinstance(data[i],float) and self.data[ID-1][adr[i]] == 'float'):
+                self.port.write(bytearray(struct.pack("f",data[i])))
+            elif(isinstance(data[i],str) and self.data[ID-1][adr[i]] == 'chr'):
+                self.port.write(data[i].encode('utf-8'))
+            elif(isinstance(data[i], int) and data[i] >= -2147483648 and data[i] <= 2147483647 and self.data[ID-1][adr[i]] == 'long'):
+                self.port.write((data[i]).to_bytes(4, byteorder='big'))
+            else:
+                raise TypeError("Transmit Error\nThe data you're trying to send does not mach the type stored in the register\nor, the pothos protocol does not support that data type\n or the data you're trying to send is too large or too small")
+        self.port.write(bytes([0xff]))
+        while(self.port.out_waiting):
+            time.sleep(0.0001)
+        acknowledge = self.port.read(4)
+        if (not(acknowledge == (bytes(b'\xff\xff\xff\xff')))):
```

Comment:
You don't need bytes() method.
```suggestion
        if (not(acknowledge == b'\xff\xff\xff\xff')):
```